File: java/src/test/java/YatzyTest.java
import com.zortek.kata.yatzy.DiceRoll;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static com.zortek.kata.yatzy.Category.*;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class YatzyTest {

    @Nested
    class Chance_should_ {
        @Test
        void scores_sum_of_all_dice() {
            //given
            final int expected = 15;

            //when
            final int actual = CHANCE.score(new DiceRoll(2, 3, 4, 5, 1));

            //then
            assertEquals(expected, actual);
        }

        @Test
            //TODO remove this unnecessary test
        void scores_sum_of_all_dice_2() {
            //given
            final int expected = 16;

            //when
            final int actual = CHANCE.score(new DiceRoll(3, 3, 4, 5, 1));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class Yatzy_should_ {
        @Test
        void scores_50_if_all_dices_are_equals() {
            //given
            final int expected = 50;

            //when
            final int actual = YATZY.score(new DiceRoll(4, 4, 4, 4, 4));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_all_dices_are_not_equals() {
            //given
            final int expected = 0;

            //when
            final int actual = YATZY.score(new DiceRoll(6, 6, 6, 6, 3));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class Ones_should_ {
        @Test
        void count_number_of_one() {
            //given
            final int expected = 1;

            //when
            final int actual = ONES.score(new DiceRoll(1, 2, 3, 4, 5));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_no_one() {
            //given
            final int expected = 0;

            //when
            final int actual = ONES.score(new DiceRoll(6, 2, 2, 4, 5));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class Twos_should_ {
        @Test
        void count_number_of_twos_multiply_by_two() {
            //given
            final int expected = 4;

            //when
            final int actual = TWOS.score(new DiceRoll(1, 2, 3, 2, 6));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_no_twos() {
            //given
            final int expected = 0;

            //when
            final int actual = TWOS.score(new DiceRoll(3, 3, 3, 3, 3));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class Threes_should_ {
        @Test
        void count_number_of_threes_multiply_by_threes() {
            //given
            final int expected = 6;

            //when
            final int actual = THREES.score(new DiceRoll(1, 2, 3, 2, 3));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_no_threes() {
            //given
            final int expected = 0;

            //when
            final int actual = THREES.score(new DiceRoll(4, 4, 4, 4, 4));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class Fours_should_ {
        @Test
        void count_number_of_fours_multiply_by_fours() {
            //given
            final int expected = 12;

            //when
            final int actual = FOURS.score(new DiceRoll(4, 4, 4, 5, 5));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_no_four() {
            //given
            final int expected = 0;

            //when
            final int actual = FOURS.score(new DiceRoll(5, 5, 5, 5, 5));

            //then
            assertEquals(expected, actual);
        }
    }


    @Nested
    class Fives_should_ {
        @Test
        void count_number_of_fives_multiply_by_fives() {
            //given
            final int expected = 10;

            //when
            final int actual = FIVES.score(new DiceRoll(4, 4, 4, 5, 5));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_no_fives() {
            //given
            final int expected = 0;

            //when
            final int actual = FIVES.score(new DiceRoll(4, 4, 4, 4, 4));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class Sixes_should_ {
        @Test
        void count_number_of_sixes_multiply_by_sixes() {
            //given
            final int expected = 18;

            //when
            final int actual = SIXES.score(new DiceRoll(6, 5, 6, 6, 5));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_no_fives() {
            //given
            final int expected = 0;

            //when
            final int actual = SIXES.score(new DiceRoll(4, 4, 4, 5, 5));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class One_pair_should_ {
        @Test
        void find_pair_and_sum_value() {
            //given
            final int expected = 6;

            //when
            final int actual = PAIR.score(new DiceRoll(3, 4, 3, 5, 6));

            //then
            assertEquals(expected, actual);
        }

        @Test
            //TODO : check function rule when we have 2 pairs
        void find_better_pair_and_sum_value() {
            //given
            final int expected = 8;

            //when
            final int actual = PAIR.score(new DiceRoll(3, 3, 4, 4, 6));

            //then
            assertEquals(expected, actual);
        }


        @Test
        void scores_0_if_no_pair() {
            //given
            final int expected = 0;

            //when
            final int actual = TWO_PAIRS.score(new DiceRoll(1, 2, 3, 4, 5));

            //then
            assertEquals(expected, actual);
        }
    }


    @Nested
    class Two_pair_should_ {
        @Test
        void find_pairs_and_sum_value() {
            //given
            final int expected = 16;

            //when
            final int actual = TWO_PAIRS.score(new DiceRoll(3, 3, 5, 4, 5));

            //then
            assertEquals(expected, actual);
        }


        @Test
        void scores_0_if_no_two_pairs() {
            //given
            final int expected = 0;

            //when
            final int actual = TWO_PAIRS.score(new DiceRoll(3, 3, 5, 4, 3));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class Three_of_a_kind_should_ {
        @Test
        void find_triple_and_sum_value() {
            //given
            final int expected = 9;

            //when
            final int actual = THREE_OF_A_KIND.score(new DiceRoll(3, 3, 3, 4, 5));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void find_triple_and_sum_value_for_quadruplet() {
            //given
            final int expected = 9;

            //when
            final int actual = THREE_OF_A_KIND.score(new DiceRoll(3, 3, 3, 3, 5));

            //then
            assertEquals(expected, actual);
        }


        @Test
        void scores_0_if_no_two_triple() {
            //given
            final int expected = 0;

            //when
            final int actual = THREE_OF_A_KIND.score(new DiceRoll(3, 1, 5, 4, 3));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class Four_of_a_kind_should_ {
        @Test
        void find_quadruplet_and_sum_value() {
            //given
            final int expected = 12;

            //when
            final int actual = FOUR_OF_A_KIND.score(new DiceRoll(3, 3, 3, 3, 5));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void find_triple_and_sum_value_for_yatzy() {
            //given
            final int expected = 12;

            //when
            final int actual = FOUR_OF_A_KIND.score(new DiceRoll(3, 3, 3, 3, 3));

            //then
            assertEquals(expected, actual);
        }


        @Test
        void scores_0_if_no_two_quadruplet() {
            //given
            final int expected = 0;

            //when
            final int actual = FOUR_OF_A_KIND.score(new DiceRoll(3, 1, 5, 4, 3));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class SmallStraight_should_ {
        @Test
        @Disabled("""
            this test won't pass before refactoring,
            Please check why
            Maybe the previous code was not compliant with the official rules https://fr.wikibooks.org/wiki/Bo%C3%AEte_%C3%A0_jeux/Yams
            """)
        void score_15_if_small_straight() {
            //given
            final int expected = 15;

            //when
            final int actual = SMALL_STRAIGHT.score(new DiceRoll(1, 2, 3, 4, 1));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void score_15_if_small_straight_US_rules() {
            //given
            final int expected = 15;

            //when
            final int actual = SMALL_STRAIGHT.score(new DiceRoll(1, 2, 3, 4, 5));

            //then
            assertEquals(expected, actual);
        }


        @Test
        void scores_0_if_no_SmallStraight() {
            //given
            final int expected = 0;

            //when
            final int actual = SMALL_STRAIGHT.score(new DiceRoll(3, 1, 5, 4, 3));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class LargeStraight_should_ {
        @Test
        @Disabled("""
            this test won't pass before refactoring
            Please check why
            Maybe the previous code was not compliant with the official rules https://fr.wikibooks.org/wiki/Bo%C3%AEte_%C3%A0_jeux/Yams
            """)
        void score_20_if_large_straight() {
            //given
            final int expected = 20;

            //when
            final int actual = LARGE_STRAIGHT.score(new DiceRoll(1, 2, 3, 4, 5));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void score_20_if_large_straight_US_Rules() {
            //given
            final int expected = 20;

            //when
            final int actual = LARGE_STRAIGHT.score(new DiceRoll(2, 3, 4, 5, 6));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_no_SmallStraight() {
            //given
            final int expected = 0;

            //when
            final int actual = LARGE_STRAIGHT.score(new DiceRoll(3, 1, 5, 4, 3));

            //then
            assertEquals(expected, actual);
        }
    }

    @Nested
    class FullHouse_should_ {
        @Test
        void sum_value_if_full_house() {
            //given
            final int expected = 18;

            //when
            final int actual = FULL_HOUSE.score(new DiceRoll(6, 2, 2, 2, 6));

            //then
            assertEquals(expected, actual);
        }

        @Test
        void scores_0_if_no_full_house() {
            //given
            final int expected = 0;

            //when
            final int actual = FULL_HOUSE.score(new DiceRoll(3, 1, 5, 4, 3));

            //then
            assertEquals(expected, actual);
        }
    }
}

File: kotlin/src/test/kotlin/YatzyTest.kt
import com.zortek.kata.yatzy.DiceRoll.Companion.of
import com.zortek.kata.yatzy.YatzyScore.*
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test


class YatzyTest {

    @Nested
    inner class `chance should` {
        @Test
        fun `scores sum of all dice`() {
            //given
            val expected = 15

            //when
            val actual = CHANCE.score(of(2, 3, 4, 5, 1))

            //then
            assertEquals(expected, actual)
        }
    }


    @Nested
    inner class `Yatzy  should  ` {
        @Test
        fun `scores  50  if  all  dices  are  equals`() {
            //given
            val expected = 50

            //when
            val actual: Int = YATZY.score(of(4, 4, 4, 4, 4))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores  0  if  all  dices  are  not  equals`() {
            //given
            val expected = 0

            //when
            val actual: Int = YATZY.score(of(6, 6, 6, 6, 3))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `Ones should ` {
        @Test
        fun `count number of one`() {
            //given
            val expected = 1

            //when
            val actual: Int = ONES.score(of(1, 2, 3, 4, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no one`() {
            //given
            val expected = 0

            //when
            val actual: Int = ONES.score(of(6, 2, 2, 4, 5))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `Twos should ` {
        @Test
        fun `count number of twos multiply by two`() {
            //given
            val expected = 4

            //when
            val actual: Int = TWOS.score(of(1, 2, 3, 2, 6))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no twos`() {
            //given
            val expected = 0

            //when
            val actual: Int = TWOS.score(of(3, 3, 3, 3, 3))

            //then
            assertEquals(expected, actual)
        }
    }


    @Nested
    inner class `Threes should ` {
        @Test
        fun `count number of threes multiply by threes`() {
            //given
            val expected = 6

            //when
            val actual: Int = THREES.score(of(1, 2, 3, 2, 3))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no threes`() {
            //given
            val expected = 0

            //when
            val actual: Int = THREES.score(of(4, 4, 4, 4, 4))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `Fours should ` {
        @Test
        fun `count number of fours multiply by fours`() {
            //given
            val expected = 12

            //when
            val actual: Int = FOURS.score(of(4, 4, 4, 5, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no four`() {
            //given
            val expected = 0

            //when
            val actual: Int = FOURS.score(of(5, 5, 5, 5, 5))

            //then
            assertEquals(expected, actual)
        }
    }


    @Nested
    inner class `Fives should ` {
        @Test
        fun `count number of fives multiply by fives`() {
            //given
            val expected = 10

            //when
            val actual: Int = FIVES.score(of(4, 4, 4, 5, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no fives`() {
            //given
            val expected = 0

            //when
            val actual: Int = FIVES.score(of(4, 4, 4, 4, 4))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `Sixes should ` {
        @Test
        fun `count number of sixes multiply by six`() {
            //given
            val expected = 18

            //when
            val actual: Int = SIXES.score(of(6, 5, 6, 6, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no six`() {
            //given
            val expected = 0

            //when
            val actual: Int = SIXES.score(of(4, 4, 4, 5, 5))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `One pair should ` {
        @Test
        fun `add pair value`() {
            //given
            val expected = 6

            //when
            val actual: Int = PAIR.score(of(3, 4, 3, 5, 6))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun  //TODO : check function rule when we have 2 pairs
                `find better pair and sum value`() {
            //given
            val expected = 8

            //when
            val actual: Int = PAIR.score(of(3, 3, 4, 4, 6))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no pair`() {
            //given
            val expected = 0

            //when
            val actual: Int = TWO_PAIRS.score(of(1, 2, 3, 4, 5))

            //then
            assertEquals(expected, actual)
        }
    }


    @Nested
    inner class `two pair should` {
        @Test
        fun `find pairs and sum value`() {
            //given
            val expected = 16

            //when
            val actual: Int = TWO_PAIRS.score(of(3, 3, 5, 4, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no two pairs`() {
            //given
            val expected = 0

            //when
            val actual: Int = TWO_PAIRS.score(of(3, 3, 5, 4, 3))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `three of a kind should` {
        @Test
        fun `find triple and sum value`() {
            //given
            val expected = 9

            //when
            val actual: Int = THREE_OF_A_KIND.score(of(3, 3, 3, 4, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `find triple and sum value for quadruplet`() {
            //given
            val expected = 9

            //when
            val actual: Int = THREE_OF_A_KIND.score(of(3, 3, 3, 3, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `score 0 if no two triple`() {
            //given
            val expected = 0

            //when
            val actual: Int = THREE_OF_A_KIND.score(of(3, 1, 5, 4, 3))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `four of a kind should` {
        @Test
        fun `find quadruplet and sum value`() {
            //given
            val expected = 12

            //when
            val actual: Int = FOUR_OF_A_KIND.score(of(3, 3, 3, 3, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `find triple and sum value for yatzy`() {
            //given
            val expected = 12

            //when
            val actual: Int = FOUR_OF_A_KIND.score(of(3, 3, 3, 3, 3))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `score 0 if no two quadruplet`() {
            //given
            val expected = 0

            //when
            val actual: Int = FOUR_OF_A_KIND.score(of(3, 1, 5, 4, 3))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `small straight should` {
        @Test
        @Disabled(
            """
            this test won't pass before refactoring
            Please check why
            Maybe the previous code was not compliant with the official rules https://fr.wikibooks.org/wiki/Bo%C3%AEte_%C3%A0_jeux/Yams
            """
        )
        fun `score 15 if small straight`() {
            //given
            val expected = 15

            //when
            val actual = SMALL_STRAIGHT.score(of(1, 2, 3, 4, 1))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `score 15 if small straight  us rules`() {
            //given
            val expected = 15

            //when
            val actual = SMALL_STRAIGHT.score(of(1, 2, 3, 4, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no small straight`() {
            //given
            val expected = 0

            //when
            val actual = SMALL_STRAIGHT.score(of(3, 1, 5, 4, 3))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `large straight should` {
        @Test
        @Disabled(
            """
            this test won't pass before refactoring
            Please check why
            Maybe the previous code was not compliant with the official rules https://fr.wikibooks.org/wiki/Bo%C3%AEte_%C3%A0_jeux/Yams
            """
        )
        fun `score 20 if large straight`() {
            //given
            val expected = 20

            //when
            val actual = LARGE_STRAIGHT.score(of(1, 2, 3, 4, 5))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 20 if large straight US rules`() {
            //given
            val expected = 20

            //when
            val actual = LARGE_STRAIGHT.score(of(2, 3, 4, 5, 6))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores 0 if no  small straight`() {
            //given
            val expected = 0

            //when
            val actual = LARGE_STRAIGHT.score(of(3, 1, 5, 4, 3))

            //then
            assertEquals(expected, actual)
        }
    }

    @Nested
    inner class `full house should` {
        @Test
        fun `sum value if full house`() {
            //given
            val expected = 18

            //when
            val actual = FULL_HOUSE.score(of(6, 2, 2, 2, 6))

            //then
            assertEquals(expected, actual)
        }

        @Test
        fun `scores0 if no full house`() {
            //given
            val expected = 0

            //when
            val actual = FULL_HOUSE.score(of(3, 1, 5, 4, 3))

            //then
            assertEquals(expected, actual)
        }
    }

}